### PR TITLE
Add decode_compressed support

### LINKED ISSUE #14887: feature request - decode_compressed
It will be great if you could add **tf.decode_compressed** to be used in **situations that tfrecords cannot be used**.

Currently, only **tf.decode_raw()** can be used in such situations, which becomes a big issue with massive amount/size of files.

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #14887 to add decode_compressed support.

The API will take a string Tensor (compressed with either ZLIB or GZIP) and output a string Tensor of the same shape with content uncompressed.

This fix fixes #14887.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/core/kernels/decode_compressed_op.cc [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    auto output_flat = output_tensor->flat<string>();
+    if (compression_type_ == "") {
+      for (int64 i = 0; i < bytes_flat.size(); i++) {
+        output_flat(i) = bytes_flat(i);
+      }
+    } else {
+      const io::ZlibCompressionOptions zlib_options =
+          compression_type_ == "ZLIB" ? io::ZlibCompressionOptions::DEFAULT()
+                                      : io::ZlibCompressionOptions::GZIP();
+      for (int64 i = 0; i < bytes_flat.size(); i++) {
```
Comment (YoelShoshan) [inline tensorflow/core/kernels/decode_compressed_op.cc]: Can you explain why do we need a loop on each byte?
Or am I reading this loop incorrectly and it's per bigger type of element?

Comment (YoelShoshan) [inline tensorflow/core/kernels/decode_compressed_op.cc]: Is this really a loop on each byte of the data?
Or is it on higher level elements?

File: tensorflow/core/ops/parsing_ops.cc [hunk 2]
```diff
@@ -48,6 +48,20 @@ output: A Tensor with one more dimension than the input `bytes`.  The
   of `bytes` divided by the number of bytes to represent `out_type`.
 )doc");
 
+REGISTER_OP("DecodeCompressed")
+    .Input("bytes: string")
+    .Output("output: string")
+    .Attr("compression_type: string = ''")
+    .SetShapeFn(shape_inference::UnchangedShape)
+    .Doc(R"doc(
+Reinterpret the bytes of a string as a vector of numbers.
```
Comment (asimshankar) [inline tensorflow/core/ops/parsing_ops.cc]: Didn't quite follow the "vector of numbers" comment here.

**Comment by martinwicke** [comment]
(from API review: This is good to go.)

File: tensorflow/core/kernels/decode_compressed_op.cc (hunk 1 above)
Comment (yongtang) [reply tensorflow/core/kernels/decode_compressed_op.cc]: @YoelShoshan It is a tensor of string so the loop at a higher level, not a byte level.

File: tensorflow/core/ops/parsing_ops.cc (hunk 2 above)
Comment (yongtang) [reply tensorflow/core/ops/parsing_ops.cc]: @asimshankar Sorry the doctoring was not corrected. The PR has been updated. Please take a look.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by yongtang** [comment]
@martinwicke Thanks for the help. It seems python 3 build still has failures (python 2 is OK). Let me take a look and update the PR.

**Comment by yongtang** [comment]
@martinwicke The Python 3 build failure has been fixed and the PR has been updated and pushed. Please take a look.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Argh sorry -- I edited the docstring to make it more explicit, could you regenerate the APIDefs?

**Comment by yongtang** [comment]
Thanks @martinwicke. The PR has been updated with `tensorflow/core/api_def/update_api_def.sh` rerun.

All tests passed except `Ubuntu Sanity`, which I think might be unrelated as I also see several other PRs (like #15491 and #15461) having the same issue.

Please take a look.

**Comment by yongtang** [comment]
@martinwicke All tests passed now but `cla/google` is stuck. Not sure how to trigger to pass the test. Is there a way to manually get around it?

**Comment by martinwicke** [comment]
Yes, that's because I added a commit. I'll merge over it.

**Comment by yongtang** [comment]
Thanks all for the help  👍 🎉 🌈

**Pull request merged**